### PR TITLE
UpSample-nearest cuda kernel update

### PR DESCRIPTION
**Pull request by jjsjann123**
updating upsampling kernel:
1. avoids atomicAdd for better fp16 performance.
2. better launch configures for 2D input.

### CODE REVIEW

**Comment by jjsjann123** [comment]
Perf number/scripts will be posted shortly.
I'll have another PR for bilinear upsampling coming soon as well.

cc @ngimel

**Comment by jjsjann123** [comment]
image

fp16 forward perf number has been observed to be all over the place, especially for tiny input. :/
Things to notice here is the speedup on the fp16 backward path.

Here's the script for the benchmark

```
import torch
import numpy as np

nrep = 300
sample_mode = "nearest"

def bench(size, fn, factor, half=False, b=2, c=32, dim=2):
   x=torch.ones([b*c*(size**dim)], device='cuda', dtype = torch.float)
   if half:
     x = x.half()
   if dim==1:
     x=x.view(b, c, size).requires_grad_()
   elif dim==2:
     x=x.view(b, c, size, size).requires_grad_()
   elif dim==3:
     x=x.view(b, c, size, size, size).requires_grad_()
   torch.cuda.synchronize()
   import time
   start = time.time()
   for i in range(nrep):
      out = fn(x, scale_factor=factor, mode=sample_mode)
   torch.cuda.synchronize()
   end = time.time()
   inp_size = x.size()
   out_size = out.size()
   del x, out
   return ((end-start)/nrep, inp_size, out_size)

def bench_back(size, fn, factor, half=False, b=2, c=32, dim=2):
   x=torch.ones([b*c*(size**dim)], device='cuda', dtype = torch.float)
   if half:
     x = x.half()
   if dim==1:
     x=x.view(b, c, size).requires_grad_()
   elif dim==2:
     x=x.view(b, c, size, size).requires_grad_()
   elif dim==3:
     x=x.view(b, c, size, size, size).requires_grad_()
   torch.cuda.synchronize()
   out = fn(x, scale_factor=factor, mode=sample_mode)
   grad = torch.randn_like(out)
   import time
   start = time.time()
   for i in range(nrep):
      out.backward(grad)
   torch.cuda.synchronize()
   end = time.time()
   inp_size = x.size()
   out_size = out.size()
   del x, out, grad
   return ((end-start)/nrep, inp_size, out_size)


spatial_size = [2**i for i in range(5,12)]
batch = [8]
channel = [32]
dim = [1, 2, 3]
scale_factor = [2]
bool_flag = [False, True]
cap = 2**31

from itertools import product
for d, b, c, s, f, half_flag in product(dim, batch, channel, spatial_size, scale_factor, bool_flag):
  if ((s*f)**d)*b*c*(2 if half_flag else 4) < cap:
    (fw_time, inp_size, out_size) = bench(s, torch.nn.functional.interpolate, f, half_flag, b, c, d)
    (bw_time, inp_size, out_size) = bench_back(s, torch.nn.functional.interpolate, f, half_flag, b, c, d)
    print(inp_size, f, half_flag, 1./fw_time, 1./bw_time)


```

**Comment by jjsjann123** [comment]
Removed the specialized 2d kernel, as the speedup is sparse. Caching seems to have done a great job saving the memory accessing pattern.

I don't think I can justify having a dedicated kernel there :cry:

**Comment by ezyang** [comment]
@ngimel happy to merge this if you give it the OK

File: aten/src/ATen/native/cuda/UpSample.cuh [hunk 1]
```diff
@@ -163,6 +163,15 @@ __device__ __forceinline__ static int nearest_neighbor_compute_source_index(
   return src_index;
 }
 
+__device__ __forceinline__ static int nearest_neighbor_compute_destination_index(
+    const float scale,
+    int src_index,
+    int output_size) {
+  const int dst_index = min<int>(
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSample.cuh]: Do you need template specialization here?

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu [hunk 2]
```diff
@@ -11,92 +11,82 @@ namespace at {
...
+  size_t dst_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (dst_idx >= dim_c * dst_dim_w)
+    return;
+
+  float scale_factor = (float)dst_dim_w / src_dim_w;
+
+  int c = (dst_idx / dst_dim_w) % dim_c;
+
+  int dst_x = dst_idx % dst_dim_w;
+  int src_x = min((int)(dst_x / scale_factor), (int)src_dim_w - 1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: Could you use `nearest_neighbor_compute_destination_index` here? Math seems the same. Even if you couldn't, `static_cast` instead of `(int)`

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu [hunk 3]
```diff
@@ -106,35 +96,33 @@ static void upsample_nearest1d_out_cuda_template(
...
   AT_ASSERT(input_width > 0 && output_width > 0);
 
+  Tensor input = input_.contiguous();
   output.resize_({input.size(0), input.size(1), output_width});
   output.zero_();
 
-  const int num_kernels = output_width;
-  const int num_threads = std::min(
-      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, 1024);
+  unsigned int n = output.numel() / nbatch;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: Is there a guard anywhere against nbatch and output.numel being 0? If not, it will lead to fpe. If there is, may be add a comment so that it's clear.

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu [hunk 4]
```diff
@@ -174,23 +162,21 @@ static void upsample_nearest1d_backward_out_cuda_template(
   grad_input.resize_({nbatch, channels, input_width});
   grad_input.zero_();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: You don't need to zero grad_input because you overwrite it?

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu [hunk 5]
```diff
@@ -174,23 +162,21 @@ static void upsample_nearest1d_backward_out_cuda_template(
...
 
-  const int num_kernels = output_width;
-  const int num_threads = std::min(
-      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, 1024);
+  unsigned int n = grad_input.numel() / nbatch;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: same comments about numel and nbatch being non-zero

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 6]
```diff
@@ -11,110 +11,114 @@ namespace at {
 namespace native {
 namespace {
 
+#define MAX_THREADS 512
+
+static int lastPow2(unsigned int n) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Why am I seeing this in so many places (last one I suspect was Thor's softmax)? Can it go into some helper header?

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 7]
```diff
@@ -11,110 +11,114 @@ namespace at {
...
-        atomicAdd(&idata[n][c][h1][w1], d2val);
+    const scalar_t* grad_o,
+    size_t dim_b,
+    size_t dim_c,
+    size_t src_dim_h,
+    size_t src_dim_w,
+    size_t dst_dim_h,
+    size_t dst_dim_w,
+    scalar_t* grad_i) {
+  assert(gridDim.y == 1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Please avoid device side asserts if possible, this can be done on the host before kernel launch

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 8]
```diff
@@ -11,110 +11,114 @@ namespace at {
...
+
+  int dst_c_stride = dst_dim_h * dst_dim_w;
+  int src_c_stride = src_dim_h * src_dim_w;
+
+  int c = (dst_idx / (dst_c_stride)) % dim_c;
+
+  float scale_factor = (float)src_dim_h / dst_dim_h;
+  int dst_y = (dst_idx / dst_dim_w) % dst_dim_h;
+  int src_y = (int)ceilf(dst_y * scale_factor);
+  int src_y_up = min((int)ceilf((dst_y + 1) * scale_factor), (int)src_dim_h);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: static_cast

File: aten/src/ATen/native/cuda/UpSample.cuh [hunk 9]
```diff
@@ -163,6 +163,15 @@ __device__ __forceinline__ static int nearest_neighbor_compute_source_index(
...
+      static_cast<int>(ceilf(src_index / scale)), output_size - 1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSample.cuh]: Wondering if precomputing invscale as 1/scale and using multiply here would speed things up?

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 10]
```diff
@@ -11,110 +11,114 @@ namespace at {
...
-      for (int c = 0; c < channels; ++c) {
-        const scalar_t val = idata[n][c][h1][w1];
-        odata[n][c][h2][w2] = val;
-      }
-    }
+  int nc_stride = blockDim.z * gridDim.z;
+
+  const size_t h1 = height1 == height2
+      ? h2
+      : nearest_neighbor_compute_source_index(height_scale, h2, height1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: yep, just do 
```
const float scale_factor = (float)height2 / (float)height1;
```
and use multiply

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 11]
```diff
@@ -11,110 +11,114 @@ namespace at {
...
+  int src_c_stride = src_dim_h * src_dim_w;
+
+  int c = (dst_idx / (dst_c_stride)) % dim_c;
+
+  float scale_factor = (float)src_dim_h / dst_dim_h;
+  int dst_y = (dst_idx / dst_dim_w) % dst_dim_h;
+  int src_y = (int)ceilf(dst_y * scale_factor);
+  int src_y_up = min((int)ceilf((dst_y + 1) * scale_factor), (int)src_dim_h);
+
+  scale_factor = (float)src_dim_w / dst_dim_w;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Sometimes you cast one operand, sometimes you cast both - is there any reason to it?

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 12]
```diff
@@ -145,26 +149,42 @@ static void upsample_nearest2d_out_cuda_template(
       input_height > 0 && input_width > 0 && output_height > 0 &&
       output_width > 0);
 
-  output.resize_({input.size(0), input.size(1), output_height, output_width});
+  Tensor input = input_.contiguous();
+  output.resize_({nbatch, channels, output_height, output_width});
   output.zero_();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Do you need .zero_() here? Output is overwritten, not accumulated, I think

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 13]
```diff
@@ -145,26 +149,42 @@ static void upsample_nearest2d_out_cuda_template(
...
-      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, 1024);
-  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  int nc = nbatch * channels;
+
+  const int max_threads = std::min<int>(
+      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, MAX_THREADS);
+
+  int block_x = std::min<int>(lastPow2(output_width), max_threads);
+  int block_y = std::min<int>(lastPow2(output_height), max_threads / block_x);
+  int block_z = std::min<int>(nc, max_threads / block_x / block_y);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Are you sure block_x, block_y, block_z can never be zero?

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu [hunk 14]
```diff
@@ -145,26 +149,42 @@ static void upsample_nearest2d_out_cuda_template(
...
+  const dim3 block(block_x, block_y, block_z);
 
+  int grid_x = cuda::ATenCeilDiv(output_width, block_x);
+  int grid_y = cuda::ATenCeilDiv(output_height, block_y);
+  int grid_z = cuda::ATenCeilDiv(nc, block_z * 4);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Guard against emtpy tensors

**Review by ngimel — CHANGES_REQUESTED** [review]
Make sure there are checks against empty tensors, and make sure you are not excessively zeroing outputs. Those already might be somewhere in the code, and I might be blind, in which case it is good to go.

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu (hunk 3 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: It's already done in the `upsample_xd_shape_check` earlier: https://github.com/pytorch/pytorch/blob/74bdcd44c4ea52896e56aef62369168abbf4a48e/aten/src/ATen/native/cuda/UpSample.cuh#L21-L124
I'll put a comment.

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu (hunk 11 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: I think Alexey's kernel casts only one operand, while the upstream code casts both. So they were sorta consistent in their own ways.

I don't think there's any perf impact on this, as they would have been casted to float any way.
I'll update to cast both just for consistency with upstream code.

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu (hunk 13 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: In this case as output_width/height != 0, no they would not be zero.
I'll put a comment here.

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu (hunk 12 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: Good catch!
it was needed for previous implementation with atomicAdd. Hence it's now obsolete. I'll remove it.

File: aten/src/ATen/native/cuda/UpSample.cuh (hunk 1 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSample.cuh]: Was blindly following the code there and I don't see any necessity for explicit specialization here neither. I'll update the other code as well.

File: aten/src/ATen/native/cuda/UpSample.cuh (hunk 9 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSample.cuh]: This is obsolete code. Will remove :)

File: aten/src/ATen/native/cuda/UpSampleNearest1d.cu (hunk 2 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest1d.cu]: Will switch to `nearest_neighbor_compute_source_index`.

File: aten/src/ATen/native/cuda/UpSampleNearest2d.cu (hunk 10 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/UpSampleNearest2d.cu]: This one is already using multiply :)

**Comment by jjsjann123** [comment]
Addressed review comments. Should be good to go :)

**Review by ngimel — APPROVED** [review]
Review comments are addressed, great job, Jie!

**Review by ezyang — APPROVED** [review]
